src/main/kotlin/com/criptext/mail/push/data/LinkDeviceActionService.kt

```diff
@@ -0,0 +1,66 @@
+package com.criptext.mail.push.data
+
+import android.app.IntentService
+import android.app.NotificationManager
+import android.content.Context
+import android.content.Intent
+
+
+class LinkDeviceActionService : IntentService("Link Device Action Service") {
+
+    companion object {
+        const val APPROVE = "Approve"
+        const val DENY = "Deny"
+
+        private var pushDataSource: PushDataSource? = null
+        private var randomId: String = ""
+        private var manager: NotificationManager? = null
+
+        fun initDataSource(dataSource: PushDataSource, id: String){
+            pushDataSource = dataSource
+            randomId = id
+            pushDataSource?.listener = dataSourceListener
+        }
+
+        private val dataSourceListener = { result: PushResult ->
+            when (result) {
+                is PushResult.LinkAccept -> onLinkAccept(result)
+                is PushResult.LinkDeny -> onLinkDeny(result)
+            }
+        }
+
+        private fun onLinkAccept(result: PushResult){
+            when(result){
+                is PushResult.LinkAccept.Success -> {
+                    manager?.cancel(result.notificationId)
+                }
+            }
+        }
+
+        private fun onLinkDeny(result: PushResult){
+            when(result){
+                is PushResult.LinkDeny.Success -> {
+                    manager?.cancel(result.notificationId)
+                }
+            }
+        }
+    }
+
+
+
+    public override fun onHandleIntent(intent: Intent?) {
+        val action = intent!!.action
+        val notificationId = intent.getIntExtra("notificationId", 0)
+        when {
+            APPROVE == action -> {
+                pushDataSource?.submitRequest(PushRequest.LinkAccept(randomId, notificationId))
```

Comment:
`IntentService` already runs in background. There is no need for `AsyncTask`, so you could just call the Worker directly.